server/pubsub/redis_query_results.go

```diff
@@ -188,20 +169,34 @@ func (r *redisQueryResults) ReadChannel(ctx context.Context, query fleet.Distrib
 	outChannel := make(chan interface{})
 	msgChannel := make(chan interface{})
 
+	conn := r.pool.Get()
+	psc := &redis.PubSubConn{Conn: conn}
+	pubSubName := pubSubForID(query.ID)
+	if err := psc.Subscribe(pubSubName); err != nil {
+		// Explicit conn.Close() here because we can't defer it until in the goroutine
+		_ = conn.Close()
+		return nil, errors.Wrapf(err, "subscribe to channel %s", pubSubName)
+	}
+
 	// Run a separate goroutine feeding redis messages into
 	// msgChannel
-	go receiveMessages(ctx, r.pool, query, msgChannel)
+	go receiveMessages(ctx, psc, msgChannel)
 
 	go func() {
+		// Unsubscribe here, but do not Close. This allows receiveMessages to finish with the final
+		// receive and non-concurrently call the Close.
+		defer psc.Unsubscribe(pubSubName)
```

Comment:
We are unsubscribing outside of the goroutine where we read, based on the same doc as before, we should keep it there?

Reply:
See above comment. I _think_ we are now using it correctly based on the docs and reading issues such as https://github.com/gomodule/redigo/issues/187.